### PR TITLE
[Workaround] Handle dangling legal files

### PR DESCRIPTION
**Pull request by keshav-space**
- Temporary workaround for https://github.com/nexB/scancode.io/issues/933

### CODE REVIEW

File: scanpipe/models.py [hunk 1]
```diff
@@ -2141,6 +2142,19 @@ def is_symlink(self):
         """Return True, if the resource is a symlink."""
         return self.type == self.Type.SYMLINK
 
+    @property
+    def is_legal(self):
```
Comment (tdruez) [inline scanpipe/models.py:2146]: Missing a unit test for this.
Also, is `or base_name.startswith(LEGAL_STARTS_ENDS)` needed? Is the start of `base_name` ever different from the start of `name`? If that's a possibility, make sure to cover it in the test.

File: scanpipe/pipes/d2d.py [hunk 2]
```diff
@@ -269,7 +269,9 @@ def scan_for_java_package(location, with_threading=True):
     return scancode._scan_resource(location, scanners, with_threading=with_threading)
 
 
-def save_java_package_scan_results(codebase_resource, scan_results, scan_errors):
+def save_java_package_scan_results(
+    codebase_resource, scan_results, scan_errors, status
```
Comment (tdruez) [inline scanpipe/pipes/d2d.py:273]: Why this change?

File: scanpipe/pipes/scancode.py [hunk 3]
```diff
@@ -231,12 +231,13 @@ def scan_for_package_data(location, with_threading=True, **kwargs):
...
 
-def save_scan_file_results(codebase_resource, scan_results, scan_errors):
+def save_scan_file_results(codebase_resource, scan_results, scan_errors, status=None):
     """
     Save the resource scan file results in the database.
     Create project errors if any occurred during the scan.
     """
-    status = flag.SCANNED
+    if not status:
+        status = flag.SCANNED
```
Comment (tdruez) [inline scanpipe/pipes/scancode.py:240]: I'm not sure about adding complexity and various status here. @pombredanne ?

**Review by tdruez — COMMENTED** [review]
Those changes add some complexity to crucial parts of the code, i.e. scanning functions, for a very specific side workaround. I don't think it is the right approach.

File: scanpipe/models.py (hunk 1 above)
Comment (keshav-space) [reply scanpipe/models.py:2146]: Added unit test.
removed `base_name.startswith(LEGAL_STARTS_ENDS)`

File: scanpipe/pipes/d2d.py (hunk 2 above)
Comment (keshav-space) [reply scanpipe/pipes/d2d.py]: fixed, we're using a separate `save_func` for legal files and no longer need to pass the status to the `save_func`.

File: scanpipe/pipes/scancode.py (hunk 3 above)
Comment (keshav-space) [reply scanpipe/pipes/scancode.py]: fixed

File: scanpipe/pipes/d2d.py [hunk 4]
```diff
@@ -1343,3 +1343,31 @@ def flag_deployed_from_resources_with_missing_license(project, doc_extensions=No
 
     no_license_files.update(status=flag.NO_LICENSES)
     unknown_license_files.update(status=flag.UNKNOWN_LICENSE)
+
+
+def handle_dangling_deployed_legal_files(project, logger):
+    to_resources = project.codebaseresources.files().to_codebase().no_status()
+    legal_files = [resource for resource in to_resources if resource.is_legal]
```
Comment (tdruez) [inline scanpipe/pipes/d2d.py:1350]: This seems highly inefficient, you should use a QuerySet.

File: scanpipe/pipes/scancode.py [hunk 5]
```diff
@@ -282,9 +282,14 @@ def scan_resources(
...
         scan_func_kwargs = {}
 
-    resource_count = resource_qs.count()
+    if isinstance(resource_qs, list):
+        resource_count = len(resource_qs)
+        resource_iterator = resource_qs
+    else:
+        resource_count = resource_qs.count()
+        resource_iterator = resource_qs.iterator(chunk_size=2000)
+
```
Comment (tdruez) [inline scanpipe/pipes/scancode.py:291]: Revert this and use a QS.

**Review by tdruez — CHANGES_REQUESTED** [review]

File: scanpipe/pipes/d2d.py (hunk 4 above)
Comment (keshav-space) [reply scanpipe/pipes/d2d.py:1350]: That's the reason it's a workaround, as of now we don't have `is_legal` as a field in our model and I don't see how we can use model property in QuerySet.

Comment (tdruez) [reply scanpipe/pipes/d2d.py:1350]: That's the point, do not use a property, create a QuerySet method instead.

Comment (keshav-space) [reply scanpipe/pipes/d2d.py]: Thanks @tdruez 
We're now using QuerySet insted property :)

File: scanpipe/pipes/scancode.py (hunk 5 above)
Comment (keshav-space) [reply scanpipe/pipes/scancode.py]: Done!

File: scanpipe/pipes/d2d.py [hunk 6]
```diff
@@ -1343,3 +1347,43 @@ def flag_deployed_from_resources_with_missing_license(project, doc_extensions=No
...
+    to `REVIEW_DANGLING_LEGAL_FILE`.
+    """
+    to_resources = project.codebaseresources.files().to_codebase().no_status()
+
+    legal_file_filter = Q()
+
+    for token in LEGAL_STARTS_ENDS:
+        legal_file_filter |= Q(name__istartswith=token)
+        legal_file_filter |= Q(name__iendswith=token)
+        legal_file_filter |= Q(name__iendswith=Concat(Value(token), F("extension")))
```
Comment (tdruez) [inline scanpipe/pipes/d2d.py:1364]: What's a data example of this case?

Comment (keshav-space) [reply scanpipe/pipes/d2d.py:1364]: Examples of these are covered in unit test.
https://github.com/nexB/scancode.io/blob/613f9fe0a951515f7adadef0715717ffa2060b70/scanpipe/tests/pipes/test_d2d.py#L1192-L1194

Comment (tdruez) [reply scanpipe/pipes/d2d.py:1364]: I don't see the `Q(name__iendswith=Concat(Value(token), F("extension")))` case.
Those are filtered using `name__iendswith` and `name__istartswith`, no?

Comment (keshav-space) [reply scanpipe/pipes/d2d.py:1364]: `project_notice.txt` would not be covered by  `name__iendswith` and `name__istartswith` since the name in itself doesn't starts or ends with token `notice`.

Comment (tdruez) [reply scanpipe/pipes/d2d.py:1364]: Fair enough, your test is missing a case such as `project.LICENSE` to cover the `Q(name__iendswith=token)`.

File: scanpipe/tests/pipes/test_d2d.py [hunk 7]
```diff
@@ -1182,3 +1182,26 @@ def test_flag_deployed_from_resources_with_missing_license(self):
...
         self.assertEqual(flag.IGNORED_DOC_FILE, from4.status)
+
+    def test_scanpipe_pipes_d2d_handle_dangling_deployed_legal_files(self):
+        to_dir = (
+            self.project1.codebase_path / "to/project.tar.zst-extract/osgi/marketplace/"
+            "resources/node_modules/foo-bar"
+        )
+        to_dir.mkdir(parents=True)
+        to_resource_files = [
+            self.data_location / "d2d/legal/poject.LICENSE",
```
Comment (tdruez) [inline scanpipe/tests/pipes/test_d2d.py:1193]:
```suggestion
            self.data_location / "d2d/legal/project.LICENSE",
```

Comment (tdruez) [reply scanpipe/tests/pipes/test_d2d.py:1193]: Typo ;)
